opentracing-metrics-micrometer/src/main/java/io/opentracing/contrib/metrics/micrometer/MicrometerMetricsReporter.java

```diff
@@ -0,0 +1,187 @@
+        Timer.Builder builder = Timer.builder(this.name).tags(tags);
+        if (publishPercentileHistogram) {
+            builder.publishPercentileHistogram();
+        }
+
+        if (null != percentiles) {
+            builder.publishPercentiles(percentiles);
+        }
+
+        if (null != sla) {
+            builder.sla(sla);
+        }
+
+        if (null != minimumExpectedValue) {
+            builder.minimumExpectedValue(minimumExpectedValue);
+        }
+
+        if (null != maximumExpectedValue) {
+            builder.maximumExpectedValue(maximumExpectedValue);
+        }
+
+        builder.register(this.registry).record(spanData.getDuration(), TimeUnit.MICROSECONDS);
+    }
+
+    public static Builder newMetricsReporter() {
+        return new Builder();
+    }
+
+    /**
+     * This builder class is responsible for creating an instance of the Micrometer
+     * metrics reporter.
+     *
+     */
+    public static class Builder {
+        private String name = "span";
+        private Duration sla, minimumExpectedValue, maximumExpectedValue;
+        private double[] percentiles;
+        private boolean publishPercentileHistogram;
+        private MeterRegistry registry = Metrics.globalRegistry;
+
+        private List<MetricLabel> metricLabels = new ArrayList<>();
+
+        public Builder withName(String name) {
+            this.name = name;
+            return this;
+        }
+
+        public Builder withCustomLabel(MetricLabel label) {
+            metricLabels.add(label);
+            return this;
+        }
+
+        public Builder withConstLabel(String name, Object value) {
+            metricLabels.add(new ConstMetricLabel(name, value));
+            return this;
+        }
+
+        public Builder withTagLabel(String name, Object defaultValue) {
+            metricLabels.add(new TagMetricLabel(name, defaultValue));
+            return this;
+        }
+
+        public Builder withBaggageLabel(String name, Object defaultValue) {
+            metricLabels.add(new BaggageMetricLabel(name, defaultValue));
+            return this;
+        }
+
+        public Builder withSla(Duration sla) {
+            this.sla = sla;
+            return this;
+        }
+
+        public Builder withMinimumExpectedValue(Duration minimumExpectedValue) {
+            this.minimumExpectedValue = minimumExpectedValue;
+            return this;
+        }
+
+        public Builder withMaximumExpectedValue(Duration maximumExpectedValue) {
+            this.maximumExpectedValue = maximumExpectedValue;
+            return this;
+        }
+
+        public Builder withPercentiles(double... percentiles) {
+            this.percentiles = percentiles;
+            return this;
+        }
+
+        public Builder enablePercentileHistogram() {
+            this.publishPercentileHistogram = true;
+            if (null == this.percentiles) {
```

Comment:
`percentiles` and `publishPercentileHistogram` are kind of orthogonal parameters as you can have one without the other. I think it is even more "correct" to pick one of the following configurations:
* set `percentiles` and `publishPercentileHistogram` to false.
* unset `percentiles` and set `publishPercentileHistogram` to true.

The latter is definitely the best approach as far as Prometheus is concerned.

AFAIU the bucket sizes of histograms are (indirectly) controlled by `minimumExpectedValue`, `maximumExpectedValue` and `sla`.

Reply:
> unset percentiles and set publishPercentileHistogram to true.

What to do when `disablePercentileHistogram()` is called then? Should `percentiles` be touched at all? I'm wondering if it's not best to just ignore `percentiles` on the `percentileHistogram` operations.

Reply:
I think `percentiles` should just be treated separately from `enable/disablePercentileHistogram` so that the default percentiles can be used, or if the user wishes, they can override them.